File: day7/day7.py
from collections import defaultdict

with open("input.txt", "r") as f:
    crabs = [int(x) for x in f.read().split(',')]

def brute(crabs):
    num_crabs = defaultdict(int)
    for crab in crabs:
        num_crabs[crab] += 1
    fuel_costs = defaultdict(int)
    possible_positions = range(1, max(crabs) + 1)
    unique_crabs = sorted(list(set(crabs)))
    for pos in possible_positions:
        fuel = 0
        for crab in unique_crabs:
            fuel += abs(crab - pos) * num_crabs[crab]
        fuel_costs[pos] += fuel
    return fuel_costs

print(min(brute(crabs).values()))

File: day7/day7.clj
(defn get_crabs [file]
  (map read-string (clojure.string/split (slurp file) #","))
)

(defn abs [n] (max n (- n)))

(defn moves [end]
  (if (<= end 1) end (+ end (moves (- end 1))))
)

(defn moves2 [end]
  (reduce + (range 1 (+ 1 end)))
)

(defn moves3 [end]
  (/ (* end (+ end 1)) 2)
)

(defn fuel_costing1 [crab pos]
  (abs (- crab pos))
)

(defn fuel_costing2 [crab pos]
  (moves3 (abs(- crab pos)))
)

  
(defn part1 [crabs]
  (for [pos (range 1 (+ (apply max crabs) 1))]
    (reduce + (map (fn [crab] (fuel_costing1 crab pos)) crabs))
  )
)

(defn part2 [crabs]
  (for [pos (range 1 (+ (apply max crabs) 1))]
    (reduce + (map (fn [crab] (fuel_costing2 crab pos)) crabs))
  )
)

(def crabs (get_crabs "input.txt"))
(println (apply min (part1 crabs)))
(println (apply min (part2 crabs)))
